File: app/src/main/java/com/irishinterest/irishinterest/network/api/irishInterest/observer/GuiObserver.java
package com.irishinterest.irishinterest.network.api.irishInterest.observer;

import java.util.HashMap;
import java.util.Map;
import java.util.Observable;
import java.util.Observer;

public class GuiObserver implements ModelToGui, GuiToModel {

    private Map<Module, GuiMessageObserver> moduleGuiMessageObserverMap;

    public GuiObserver() {
        moduleGuiMessageObserverMap = new HashMap<>();
        Module[] modules = Module.values();
        for (Module module : modules) {
            moduleGuiMessageObserverMap.put(module, new GuiMessageObserver());
        }
    }

    @Override
    public void registerObserver(Module module, Observer o) {
        moduleGuiMessageObserverMap.get(module).addObserver(o);
    }

    @Override
    public void unregisterObserver(Module module, Observer o) {
        moduleGuiMessageObserverMap.get(module).deleteObserver(o);
    }


    @Override
    public void registerObserverModel(Module module, Observer o) {
        moduleGuiMessageObserverMap.get(module).addObserver(o);
    }

    @Override
    public void unregisterObserverModel(Module module, Observer o) {
        moduleGuiMessageObserverMap.get(module).deleteObserver(o);
    }

    @Override
    public void publishMessage(GuiMessage guiMessage) {
        moduleGuiMessageObserverMap.get(guiMessage.getModule()).publishMessage(guiMessage);
    }

    @Override
    public void publishMessageModel(GuiMessage guiMessage) {
        moduleGuiMessageObserverMap.get(guiMessage.getModule()).publishMessage(guiMessage);
    }

    private class GuiMessageObserver extends Observable {
        void publishMessage(GuiMessage guiMessage) {
            setChanged();
            notifyObservers(guiMessage);
        }
    }
}

File: app/src/main/java/com/irishinterest/irishinterest/network/api/irishInterest/observer/GuiObserver.kt
package com.irishinterest.irishinterest.network.api.irishInterest.observer

import com.irishinterest.irishinterest.network.api.irishInterest.observer.ModelToGui
import com.irishinterest.irishinterest.network.api.irishInterest.observer.GuiToModel
import com.irishinterest.irishinterest.network.api.irishInterest.observer.GuiObserver.GuiMessageObserver
import com.irishinterest.irishinterest.network.api.irishInterest.observer.GuiMessage
import java.util.*

class GuiObserver : ModelToGui, GuiToModel {
    private val moduleGuiMessageObserverMap: MutableMap<Module, GuiMessageObserver>
    override fun registerObserver(module: Module, o: Observer) {
        moduleGuiMessageObserverMap[module]!!.addObserver(o)
    }

    override fun unregisterObserver(module: Module, o: Observer) {
        moduleGuiMessageObserverMap[module]!!.deleteObserver(o)
    }

    override fun registerObserverModel(module: Module, o: Observer) {
        moduleGuiMessageObserverMap[module]!!.addObserver(o)
    }

    override fun unregisterObserverModel(module: Module, o: Observer) {
        moduleGuiMessageObserverMap[module]!!.deleteObserver(o)
    }

    override fun publishMessage(guiMessage: GuiMessage<*>) {
        moduleGuiMessageObserverMap[guiMessage.module]!!.publishMessage(guiMessage)
    }

    override fun publishMessageModel(guiMessage: GuiMessage<*>) {
        moduleGuiMessageObserverMap[guiMessage.module]!!.publishMessage(guiMessage)
    }

    private inner class GuiMessageObserver : Observable() {
        fun publishMessage(guiMessage: GuiMessage<*>?) {
            setChanged()
            notifyObservers(guiMessage)
        }
    }

    init {
        moduleGuiMessageObserverMap = HashMap()
        val modules = Module.values()
        for (module in modules) {
            moduleGuiMessageObserverMap[module] = GuiMessageObserver()
        }
    }
}